meshpy/geometry_set.py

```diff
@@ -70,41 +70,53 @@ def __init__(self, geometry_type, name=None, **kwargs):
         self.geometry_type = geometry_type
         self.name = name
 
-    def link_to_nodes(self):
+    def link_to_nodes(self, *, link_to_nodes="explicitly_contained_nodes"):
         """Set a link to this object in the all contained nodes of this geometry set.
-        This will also set a link to nodes connected to an edge in a node based line
-        set.
+
+        link_to_nodes: str
+            "explicitly_contained_nodes":
+                A link will be set for all nodes that are explicitly part of the geometry set
+            "all_nodes":
+                A link will be set for all nodes that are part of the geometry set, i.e., also
+                nodes connected to elements of an element set. This is mainly used for vtk
+                output so we can color the nodes which are part of element sets.
         """
-        for node in self.get_all_nodes():
+        if link_to_nodes == "explicitly_contained_nodes":
+            node_list = self.get_node_dict().keys()
+        elif link_to_nodes == "all_nodes":
+            node_list = self.get_all_nodes()
+        else:
+            raise ValueError(f'Got unexpected value link nodes="{link_to_nodes}"')
+        for node in node_list:
             node.node_sets_link.append(self)
 
     def check_replaced_nodes(self):
+        """Check if nodes in this set have to be replaced.
+        We need to do this for explicitly contained nodes in this set.
         """
-        Check if nodes in this set have to be replaced.
-        We need to do this for all nodes to correctly represent node-based sets.
-        """
-        for node in self.get_all_nodes():
+        # Don't iterate directly over the keys as the dict changes during this iteration
+        for node in list(self.get_node_dict().keys()):
             if node.master_node is not None:
                 self.replace_node(node, node.get_master_node())
 
     def replace_node(self, old_node, new_node):
-        """Replace old_node with new_node. This is only done for point sets."""
-
-        # Check if the new node is in the set.
-        my_nodes = self.get_all_nodes()
-        has_new_node = new_node in my_nodes
-
-        for i, node in enumerate(my_nodes):
-            if node == old_node:
-                if has_new_node:
-                    del my_nodes[i]
-                else:
-                    my_nodes[i] = new_node
-                break
-        else:
-            raise ValueError(
-                "The node that should be replaced is not in the current node set"
-            )
+        """Replace old_node with new_node."""
+
+        explicit_nodes_in_this_set = self.get_node_dict()
+        explicit_nodes_in_this_set[new_node] = None
+        del explicit_nodes_in_this_set[old_node]
```

Comment:
I've actually never seen this `del` syntax.

Maybe

```suggestion
        explicit_nodes_in_this_set.pop(old_node)
```

This is more pythonic/expressive

Reply:
I would keep it as it is, as `del` does exactly what is expected here, `pop` would additionally return the value of the deleted key, which is not needed here. See also [this thread](https://stackoverflow.com/questions/5713218/what-is-difference-between-pop-vs-del-for-deleting-a-dictionary-item-python).